interop-testing/src/test/java/io/grpc/testing/integration/CascadingTest.java

```diff
@@ -0,0 +1,334 @@
+      // receive cancellation.
+      assertEquals(Status.Code.DEADLINE_EXCEEDED, status.getCode());
+      // Count the outermost failure as cancellation
+      observedCancellations.incrementAndGet();
+
+      assertEquals(observedCancellations.get(), depth.get());
+      assertEquals(observedCancellations.get(), receivedCancellations.get());
+      assertEquals(3, depth.get());
+    }
+  }
+
+  /**
+   * Test that when RPC cancellation propagates up a call chain, the cancellation of the parent
+   * RPC triggers cancellation of all of its children.
+   */
+  @Test
+  public void testCascadingCancellationViaLeafFailure() throws Exception {
+    startCallTreeServer();
+    try {
+      // Use response size limit to control tree depth.
+      blockingStub.unaryCall(Messages.SimpleRequest.newBuilder().setResponseSize(3).build());
+      fail("Expected abort");
+    } catch (StatusRuntimeException sre) {
+      // Wait for the workers to finish
+      Thread.sleep(100);
+      Status status = Status.fromThrowable(sre);
+      // Outermost caller observes ABORTED propagating up from the failing leaf,
+      // The descendant RPCs are cancelled so they receive CANCELLED.
+      assertEquals(Status.Code.ABORTED, status.getCode());
+      // All nodes (15) except one edge of the tree (4) will be cancelled.
+      assertEquals(11, observedCancellations.get());
+      assertEquals(11, receivedCancellations.get());
+    }
+  }
+
+  /**
+   * Create a chain of client to server calls which can be cancelled top down.
+   */
+  private void startChainingServer() throws IOException {
+    final Executor otherWork = Context.propagate(Executors.newCachedThreadPool());
+    server = InProcessServerBuilder.forName("channel").addService(
+        ServerInterceptors.intercept(TestServiceGrpc.bindService(service),
+            new ServerInterceptor() {
+              @Override
+              public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+                  MethodDescriptor<ReqT, RespT> method,
+                  final ServerCall<RespT> call,
+                  Metadata headers,
+                  ServerCallHandler<ReqT, RespT> next) {
+                // Respond with the headers but nothing else.
+                call.sendHeaders(new Metadata());
+                call.request(1);
+                return new ServerCall.Listener<ReqT>() {
+                  @Override
+                  public void onMessage(final ReqT message) {
+                    try {
+                      // Wait and then recurse.
+                      if (depth.incrementAndGet() == depthThreshold) {
+                        call.close(Status.ABORTED, new Metadata());
+                        return;
+                      }
+                      Thread.sleep(50);
+                    } catch (InterruptedException ie) {
+                      throw new RuntimeException(ie);
+                    }
+
+                    otherWork.execute(new Runnable() {
```

Comment:
I guess you are trying to make this more "realistic" by going through the executor?